extras/findkey/findkey.1
.TH FINDKEY 1 
.ad
.fi
.SH NAME
findkey
\-
find high-entropy data
.SH SYNOPSIS
.na
.nf
.ad
.fi
\fBfindkey\fR [\fB-cdlmx\fR] [\fB-b \fIbias\fR] [\fB-w \fIwindow\fR]
[\fB-t \fIthreshold\fR] \fIfile...\fR
.SH DESCRIPTION
.ad
.fi
\fBfindkey\fR moves a fixed-size window over the named files (stdin
by default) and reports window positions containing a large number
of different symbols. This approach, described by Shamir and van
Someren, simplifies Shannon's entropy formula and allows for faster
searching.

By default, overlapping regions of \fIwindow\fR bytes with
high-entropy content are reported as one large region:

.ti +4
\fIfile start [content] length\fR

Fields are tab separated. By default no content is shown.

In order to display the content of potential keys, specify the
\fB-x\fR option (show content).

In order to report potential keys individually, specify the
\fB-d\fR option (detail mode).

Options:
.IP \fB-c\fR
Calibration mode. Report each high-entropy region of \fIwindow\fR
bytes as: \fIfile start threshold bias [content]\fR. This is useful
for deriving good \fIthreshold\fR and \fIbias\fR values from real
data. Use of this option implies \fB-d\fR.
.IP \fB-d\fR
Detail mode. Report each high-entropy region of \fIwindow\fR bytes
as: \fIfile start [content]\fR. This option increases the amount
of output considerably. Combine this option with with \fB-x\fR to
display the content of individual keys.
.IP "\fB-b \fIbias\fR (default: 31)"
Exclude data that is too monotonic. \fBfindkey\fR counts how often
successive byte values increase or decrease. The number of increments
and decrements in a window must not differ by more than \fIbias\fR.
For 64-byte windows, 22 is a good bias value.
.IP \fB-l\fR
Report only the names of files containing high-entropy regions.
.IP "\fB-m\fR (default)"
Merge mode. Report multiple overlapping high-entropy windows as
one region, using the format: \fIfile start [content] length\fR.
Use of this option can reduce the amount of output considerably.
.IP "\fB-t \fIthreshold\fR (default: 95)"
Search for windows with at least \fIthreshold\fR different symbols.
For 64-byte windows, 52 is a good threshold value.
.IP "\fB-w \fIwindow\fR (default: 128)"
Search for high-entropy regions of at least \fIwindow\fR bytes.
Use of \fB-w\fR also requires use of the \fB-b\fR and \fB-t\fR options.
.IP \fB-x\fR
Display high-entropy content in hexadecimal form.
.SH HISTORY
.na
.nf
.fi
.ad
This \fBfindkey\fR command was written for the coroner's toolkit.
.SH LICENSE
.na
.nf
This software is distributed under the IBM Public License.
.SH SEE ALSO
.na
.nf
.ad
.fi
A. Shamir and N. van Someren, Playing Hide and Seek With Stored
Keys, 1998.
.nf
.na
http://www.ncipher.com/products/files/papers/anguilla/keyhide2.pdf.
.PP
.ad
.fi
C.E.  Shannon, A mathematical theory of communication. Bell
System Technical Journal 27 (1948) 379-423, 632-656.
.nf
.na
http://cm.bell-labs.com/cm/ms/what/shannonday/shannon1948.ps.gz
.SH AUTHOR(S)
.na
.nf
Wietse Venema
IBM T.J. Watson Research
P.O. Box 704
Yorktown Heights, NY 10598, USA


CHANGES
Sat Aug 29 2009

o	grave-robber did not record the initial directory in the body file.

o	Bit-rot: FreeBSD 7 support, including ``struct acct'' changes.

o	Bit-rot: lastcomm format strings.

Sun Jan 11 20:01:27 EST 2009

o	Bit-rot: OpenBSD 4.x support (still TODO: file system tools).

Thu Jun  7 10:54:44 EDT 2007

o	Bit-rot: FreeBSD 6.x support.

Mon Apr  9 09:36:31 EDT 2007

o	Bit-rot: the 64-bit workaround for LINUX lseek() is no
	longer needed, and it didn't compile on some systems.

Thu Jan 11 16:10:29 EST 2007

o	Bugfix: the timeout command always reported zero exit status
	after voluntary child process exit.

Mon Jun 26 18:44:17 EDT 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Fri Mar 17 09:25:57 EST 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Tue Jan  6 12:25:32 EST 2004

o       Workaround: mactime treated "," as a special character in
	file names, now it's 0xff.

o	Support for ext[23]fs inode file sizes > 32bits.

o	Bugfix: ils file sizes larger than unsigned long.

Tue Oct 14 18:23:49 EDT 2003

o       pcat now fully supports FreeBSD systems that do not have
	/proc mounted (it's no longer mounted by default with
	FreeBSD 5). pcat tries to use /proc if it can, and uses
	ptrace() and gropes kernel memory if it has to.  Operation
	without /proc requires super-user privileges.

Thu Oct  2 09:14:00 EDT 2003

o       Completed support for UFS1 and UFS2 in FreeBSD 5.x. However,
	the pcat command still requires that the /proc file system
	is mounted. This will be fixed later.

Sat Aug 30 19:21:29 EDT 2003

o	Preliminary FreeBSD 5.0 port. UFS2 support is still to be done.

Fri Aug  1 10:07:39 EDT 2003

o	RedHat 9.0 fix by Florin Andrei for the file command.

Sat Oct  5 13:47:29 EDT 2002

o	Restored tctutils compatibility, which broke with 20020916.

Mon Sep 16 14:39:09 EDT 2002

o       The lowest inode examined is now the lowest possible inode
	number instead of ROOTINUM.

Fri Mar 22 14:41:32 EST 2002

o	Added support for OpenBSD 3.0.

Wed Dec 12 13:21:00 EST 2001

o	tm_print() and tm_string() did unnecessary %HEX expansion
	of $ and ^ characters (Philippe Bourgeois).

o       mactime mis-sorted dates before/after early September 2001
	(Philippe Bourgeois, Brian Carrier).

Thu Oct  4 15:48:42 EDT 2001

o       Linux can handle files > 2Gbyte, but only if we specify
	extra compiler flags.

o	Solaris 8/64bit needs an extra include file in pcat.c.

Sun Sep  9 17:23:13 EDT 2001

o       Discovered to my horror that unrm on ext2fs can have an
	off-by-one error when testing a bit flag in the data block
	allocation bitmaps - and this problem is dependent on the
	Linux version being used.

o       Made the timeout command a little more verbose so it is
	less of a mystery when the child process is killed.

Thu Sep  6 13:51:00 EDT 2001

o	New grave-robber -I option for better control of icat/pcat.

Mon May 21 10:11:19 EDT 2001

o       findkey command-line error check was too conservative,
	and disallowed threshold of 256 different byte values
	(Ovanes Manucharyan).

Mon May 14 13:11:54 EDT 2001

o	Patch 06 broke lib/pass.cache.pl, which broke mactime.

o	Patch 06 did not update the patchlevel file.

Fri Mar 16 14:25:11 EST 2001

o	Fixed rpm root directory when running on a corpse.
	Marco Walter (SUN).

o	Fixed file(1) output parsing problem with multiple :
	characters.  Marco Walter (SUN)

Thu Feb 22 11:16:40 EST 2001

o	Patch by Brian Carrier (Purdue) so that fs_read_random()
	properly maintains file offset information.

Thu Feb  8 10:40:02 EST 2001

o	grave-robber dies silently when no group file is found.
	This bites when examining a partial machine corpse
	(bin/grave-robber). Klaus-Peter Kossakowski.

o	The "do first looks" code did not relocate to $TCT_HOME
	(lib/tree.pl). Klaus-Peter Kossakowski.

Tue Jan 23 18:44:36 EST 2001

o       The grave-robber did not correctly handle multi-line df
	output records (ps_spy.pl, suck_free_inodes.pl).

o       The grave-robber mounted corpse inode scan looked only at
	the corpse's "root" file system (suck_free_inodes.pl).

Mon Jan 22 11:23:32 EST 2001

o       grave-robber enabled the default -lPO flags when -c was
	specified, causing the program to gripe and exit.
	
Tue Sep 26 10:55:24 EDT 2000

o       Fixed a bug in fs_read_block() that caused icat etc. to
	remember the wrong file seek position, so that icat etc.
	would call lseek() too often. The other possibility, not
	calling lseek() when it should be called, did not happen
	due to an implementation artefact of the TCT utilities.
	Reported by Brian Carrier, Purdue University.

	Updated help-recovering-file document by Dan.

Sat Aug 12 10:51:16 EDT 2000

o       Added -V option to direct icat/ils/unrm for verbose output
	to stdout.  The -v option now sends verbose logging to stderr.

Tue Aug  8 11:46:19 PDT 2000

o	created "help-recovering-file", added note to README.FIRST

Sat Aug  5 11:34:26 PDT 2000

o	mactime time range repaired, fixed docs, -l option slightly
	fixed... perhaps deprecation in the future?

Fri Aug  4 20:06:15 EDT 2000

o	ils and ils2mac now include device names for more
	readable mactime reports of dead inodes.

o	mactime could not run with current working directory
	different from $TCT_HOME.

o	mactime -d could destroy an existing $DATA/body file.

o	mactime -d turned on verbosity, instead of letting -v do that.

o	mactime -s (suid) flag used an obsolete method of getting data.

o	ils2mac broke because perl split() can't see the difference
	between empty last fields and missing last fields.

o	mactime mistakenly skipped files with all-zero time stamps.

o	ils2mac now emits ls -l like info for mactime compatibility.

Tue Aug  1 13:50:20 PDT 2000

o	Prepend our directory list to @INC rather than clobbering it.

Mon Jul 31 17:37:35 PDT 2000

o	removed mention of sans.org from help-when-broken-into


extras/findkey/findkey.c
/*++
/* NAME
/*	findkey 1
/* SUMMARY
/*	find high-entropy data
/* SYNOPSIS
/* .ad
/* .fi
/*	\fBfindkey\fR [\fB-cdlmx\fR] [\fB-b \fIbias\fR] [\fB-w \fIwindow\fR]
/*	[\fB-t \fIthreshold\fR] \fIfile...\fR
/* DESCRIPTION
/*	\fBfindkey\fR moves a fixed-size window over the named files (stdin
/*	by default) and reports window positions containing a large number
/*	of different symbols. This approach, described by Shamir and van
/*	Someren, simplifies Shannon's entropy formula and allows for faster
/*	searching.
/*
/*	By default, overlapping regions of \fIwindow\fR bytes with 
/*	high-entropy content are reported as one large region:
/*
/* .ti +4
/*	\fIfile start [content] length\fR
/*
/*	Fields are tab separated. By default no content is shown.
/*
/*	In order to display the content of potential keys, specify the
/*	\fB-x\fR option (show content).
/*
/*	In order to report potential keys individually, specify the 
/*	\fB-d\fR option (detail mode).
/*
/*	 Options:
/* .IP \fB-c\fR
/*	Calibration mode. Report each high-entropy region of \fIwindow\fR 
/*	bytes as: \fIfile start threshold bias [content]\fR. This is useful
/*	for deriving good \fIthreshold\fR and \fIbias\fR values from real
/*	data. Use of this option implies \fB-d\fR.
/* .IP \fB-d\fR
/*	Detail mode. Report each high-entropy region of \fIwindow\fR bytes
/*	as: \fIfile start [content]\fR. This option increases the amount
/*	of output considerably. Combine this option with with \fB-x\fR to 
/*	display the content of individual keys.
/* .IP "\fB-b \fIbias\fR (default: 31)"
/*	Exclude data that is too monotonic. \fBfindkey\fR counts how often
/*	successive byte values increase or decrease. The number of increments
/*	and decrements in a window must not differ by more than \fIbias\fR.
/*	For 64-byte windows, 22 is a good bias value.
/* .IP \fB-l\fR
/*	Report only the names of files containing high-entropy regions.
/* .IP "\fB-m\fR (default)"
/*	Merge mode. Report multiple overlapping high-entropy windows as
/*	one region, using the format: \fIfile start [content] length\fR.
/*	Use of this option can reduce the amount of output considerably.
/* .IP "\fB-t \fIthreshold\fR (default: 95)"
/*	Search for windows with at least \fIthreshold\fR different symbols.
/*	For 64-byte windows, 52 is a good threshold value.
/* .IP "\fB-w \fIwindow\fR (default: 128)"
/*	Search for high-entropy regions of at least \fIwindow\fR bytes.
/*	Use of \fB-w\fR also requires use of the \fB-b\fR and \fB-t\fR options.
/* .IP \fB-x\fR
/*	Display high-entropy content in hexadecimal form.
/* HISTORY
/* .fi
/* .ad
/*	This \fBfindkey\fR command was written for the coroner's toolkit.
/* LICENSE
/*	This software is distributed under the IBM Public License.
/* SEE ALSO
/* .ad
/* .fi
/*	A. Shamir and N. van Someren, Playing Hide and Seek With Stored
/*	Keys, 1998.
/* .nf
/* .na
/*	http://www.ncipher.com/products/files/papers/anguilla/keyhide2.pdf.
/* .PP
/* .ad
/* .fi
/*      C.E.  Shannon, A mathematical theory of communication. Bell
/*	System Technical Journal 27 (1948) 379-423, 632-656.
/* .nf
/* .na
/*	http://cm.bell-labs.com/cm/ms/what/shannonday/shannon1948.ps.gz
/* AUTHOR(S)
/*	Wietse Venema
/*	IBM T.J. Watson Research
/*	P.O. Box 704
/*	Yorktown Heights, NY 10598, USA
/*--*/

/* System library. */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <errno.h>

/* SunOS4 needs this - I hope it does not break on some pedantic systems. */

extern char *optarg;
extern int optind;

/* Utility library. */

#include <error.h>

/* Application-specific. */

#define ALL_BYTES	256
#define DEF_WINDOW_SIZE	128		/* 64 for exportable keys */
#define DEF_THRESHOLD	95		/* 52 for exportable keys */
#define DEF_BIAS	31		/* 22 for exportable keys */

#define FLAG_DETAIL	(1<<0)		/* report individual windows */
#define FLAG_LIST	(1<<1)		/* list file names only */
#define FLAG_HEXDUMP	(1<<2)		/* list content */
#define FLAG_CALIBR	(1<<3)		/* calibration mode */

/* search_he - search stream for high-entry sequence */

void    search_he(char *path, FILE *fp, int window_size, int threshold, int bias, int flags)
{
    unsigned char ring_buf[window_size];
    int     byte_frequency[ALL_BYTES];
    int     distinct_bytes = 0;
    unsigned file_offset = 0;
    int     ring_offset;
    int     ch;
    int     i;
    int     high_entropy = 0;
    unsigned start;
    unsigned end;
    int     last = 0;
    int     direction = 0;
    int     dir_buf[window_size];

    /*
     * Initialize.
     */
    for (i = 0; i < ALL_BYTES; i++)
	byte_frequency[i] = 0;
    for (i = 0; i < window_size; i++)
	dir_buf[i] = 0;

    /*
     * A short and sweet loop that slides the window over the file and that
     * on the fly maintains symbol frequency counts and distinct symbol
     * counts.
     */
    while ((ch = getc(fp)) != EOF) {
	ring_offset = file_offset % window_size;
	if (file_offset >= window_size)
	    if (--byte_frequency[ring_buf[ring_offset]] == 0)
		distinct_bytes -= 1;
	ring_buf[ring_offset] = ch;
	if (byte_frequency[ch]++ == 0)
	    distinct_bytes += 1;
	file_offset++;
	direction -= dir_buf[ring_offset];
	direction += dir_buf[ring_offset] =
	    (ch < last ? -1 : ch == last ? 0 : +1);
	last = ch;
	if (file_offset < window_size)
	    continue;

	/*
	 * A lot of code to report what we found.
	 */
	if (distinct_bytes >= threshold
	    && direction <= bias
	    && direction >= -bias) {

	    /*
	     * Report each high-entropy window by itself. This can produce an
	     * incredible amount of output, but is useful for tuning and for
	     * reporting individual keys.
	     */
	    if (flags & FLAG_DETAIL) {
		printf("%s\t%u", path, file_offset - window_size);
		if (flags & FLAG_CALIBR)
		    printf("\t%u\t%d", distinct_bytes,
			   direction < 0 ? -direction : direction);
		if (flags & FLAG_HEXDUMP) {
		    printf("\t");
		    for (i = 0; i < window_size; i++)
			printf("%02x", ring_buf[(file_offset + i) % window_size]);
		}
		printf("\n");
		continue;
	    }

	    /*
	     * Report file names only.
	     */
	    if ((flags & FLAG_LIST) != 0) {
		printf("%s\n", path);
		break;
	    }

	    /*
	     * Concatenate overlapping high-entropy windows. This reduces the
	     * amount of output considerably.
	     */
	    if (high_entropy == 0) {
		high_entropy = 1;
		start = file_offset - window_size;
		printf("%s\t%u\t", path, start);
		if (flags & FLAG_HEXDUMP)
		    for (i = 0; i < window_size; i++)
			printf("%02x", ring_buf[(file_offset + i) % window_size]);
	    } else if (flags & FLAG_HEXDUMP)
		printf("%02x", ch);
	    end = file_offset;

	} else {

	    if (flags & FLAG_DETAIL)
		continue;

	    /*
	     * Concatenate overlapping high-entropy windows. This reduces the
	     * amount of output considerably.
	     */
	    if (high_entropy != 0 && file_offset - window_size >= end) {
		high_entropy = 0;
		if (flags & FLAG_HEXDUMP)
		    printf("\t");
		printf("%u\n", end - start);
	    }
	}
    }
    if (high_entropy != 0) {
	if (flags & FLAG_HEXDUMP)
	    printf("\t");
	printf("%u\n", file_offset - start);
    }
}

/* main - driver */

int     main(int argc, char **argv)
{
    FILE   *fp;
    int     window_size = DEF_WINDOW_SIZE;
    int     threshold = DEF_THRESHOLD;
    int     bias = DEF_BIAS;
    int     flags = 0;
    int     ch;
    int     i;

    progname = argv[0];

    /*
     * Parse JCL.
     */
    while ((ch = getopt(argc, argv, "b:cdlmt:w:x")) > 0) {
	switch (ch) {
	case 'c':
	    flags |= (FLAG_DETAIL | FLAG_CALIBR);
	    break;
	case 'd':
	    flags |= FLAG_DETAIL;
	    break;
	case 'b':
	    if ((bias = atoi(optarg)) <= 0)
		error("bad bias value: %s", optarg);
	    break;
	case 'l':
	    flags |= FLAG_LIST;
	    break;
	case 'm':
	    flags &= ~(FLAG_DETAIL | FLAG_CALIBR);
	    break;
	case 't':
	    if ((threshold = atoi(optarg)) <= 0)
		error("bad threshold value: %s", optarg);
	    if (threshold > ALL_BYTES)
		error("threshold cannot be larger than %d", ALL_BYTES);
	    break;
	case 'w':
	    if ((window_size = atoi(optarg)) <= 0)
		error("bad window size: %s", optarg);
	    break;
	case 'x':
	    flags |= FLAG_HEXDUMP;
	    break;
	default:
	    error("usage: %s [-b bias] [-c (calibrate)] [-d (detail)] [-l (names only)] [-m (merge windows)] [-t threshold] [-w window_size] [-x (hexdump)] [file...]",
		  progname);
	}
    }

    /*
     * Sanity checks.
     */
    if (threshold > window_size)
	error("threshold %d exceeds window size %d", threshold, window_size);
    if (bias > window_size)
	remark("bias cannot exceed window_size");

    /*
     * Search for keys.
     */
    if (argc == optind) {
	search_he("(stdin)", stdin, window_size, threshold, bias, flags);
    } else {
	for (i = optind; i < argc; i++) {
	    if ((fp = fopen(argv[i], "r")) == 0) {
		remark("open %s: %m", argv[i]);
	    } else {
		search_he(argv[i], fp, window_size, threshold, bias, flags);
		if (fclose(fp) != 0)
		    remark("read %s: %m\n", argv[i]);
	    }
	}
    }
    exit(0);
}
